src/loader/mod.rs

```diff
@@ -0,0 +1,580 @@
+            Error::InvalidEntryAddress => "Invalid entry address",
+            Error::InvalidBzImage => "Invalid bzImage",
+            Error::InvalidKernelStartAddress => "Invalid kernel start address",
+            Error::InitrdImageSizeTooLarge => "Initrd image size too large",
+            Error::ReadElfHeader => "Unable to read elf header",
+            Error::ReadKernelImage => "Unable to read kernel image",
+            Error::ReadProgramHeader => "Unable to read program header",
+            Error::ReadBzImageHeader => "Unable to read bzImage header",
+            Error::ReadBzImageCompressedKernel => "Unable to read bzImage compressed kernel",
+            Error::ReadInitrdImage => "Unable to read initrd image",
+            Error::SeekKernelStart => "Unable to seek to kernel start",
+            Error::SeekElfStart => "Unable to seek to elf start",
+            Error::SeekProgramHeader => "Unable to seek to program header",
+            Error::SeekBzImageEnd => "Unable to seek bzImage end",
+            Error::SeekBzImageHeader => "Unable to seek bzImage header",
+            Error::SeekBzImageCompressedKernel => "Unable to seek bzImage compressed kernel",
+            Error::SeekInitrdImage => "Unable to seek initrd image",
+        }
+    }
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Kernel Loader Error: {}", Error::description(self))
+    }
+}
+
+/// * `kernel_load` - The actual `guest_mem` address where kernel image is loaded start.
+/// * `kernel_end` - The offset of `guest_mem` where kernel image load is loaded finish, return
+///                  in case of loading initrd adjacent to kernel image.
+/// * `setup_header` - The setup_header belongs to linux boot protocol, only for bzImage, vmm
+///                    will use it to setup setup_header.init_size, which is a must for bzImage
+///                    direct boot.
+#[derive(Debug, Default, Copy, Clone, PartialEq)]
+pub struct KernelLoaderResult {
+    pub kernel_load: GuestAddress,
+    pub kernel_end: GuestUsize,
+    pub setup_header: Option<bootparam::setup_header>,
+}
+
+pub trait KernelLoader {
+    fn load<F>(
+        guest_mem: &GuestMemoryMmap,
+        kernel_start: Option<GuestAddress>,
+        kernel_image: &mut F,
+        lowest_kernel_start: Option<GuestAddress>,
+    ) -> Result<(KernelLoaderResult)>
+    where
+        F: Read + Seek;
+}
+
+pub struct Elf;
+
+impl KernelLoader for Elf {
+    /// Loads a kernel from a vmlinux elf image to a slice
+    ///
+    /// kernel is loaded into guest memory at offset phdr.p_paddr specified by elf image.
+    ///
+    /// # Arguments
+    ///
+    /// * `guest_mem` - The guest memory region the kernel is written to.
+    /// * `kernel_start` - The offset into 'guest _mem' at which to load the kernel.
+    /// * `kernel_image` - Input vmlinux image.
+    /// * `lowest_kernel_start` - This is the start of the high memory, kernel should above it.
```

Comment:
Shouldn't this be something like `highmem_start_address` or just `start_address` since it is sort of implied that the kernel is loaded in the high memory? I personally prefer have the name of the variables suffixed with what they actually represent when it is not clear from the name (in this case I would use address).

Reply:
Thanks! I prefer to highmem_start_address :-)